File: console/shared/lib/catalog/clashroyale.ts
// Copyright (c) 2026 Adam Ousmer. All rights reserved.
// Proprietary. No license granted. See LICENSE.md.

import type { ModuleDef } from './module-def';

export const CLASHROYALE_MODULE: ModuleDef = 
{
  // All four !cr views ride gossip's one shared Clash Royale profile cache,
  // so their token palettes below mirror app/sesame/modules/clashroyale.go.
  id: 'clashroyale',
  label: 'Clash Royale Stats',
  tagline: 'Clash Royale profiles, decks and Path of Legends standing in chat.',
  description:
    'One command, four looks: !cr shows a player\'s lifetime profile (level, win/loss record, win rate, three-crown wins, clan); !cr decks lists their current battle deck with the average elixir cost; !cr ranked shows their Path of Legends standing (falling back to legacy league seasons for older accounts); !cr road shows their trophy-road record and arena. The squashed forms !crstats, !crdecks, !crranked and !crroad work too. Link your player tag below — Clash Royale has no name lookup, so a tag like #P2LQ0GR is required. Viewers can also name any tag, e.g. "!cr #P2LQ0GR".',
  icon: 'crown',
  category: 'Stats',
  defaultEnabled: false,
  replies: [
    {
      key: 'stats',
      label: '!cr',
      tagline: 'Lifetime profile — !cr, !cr stats or !crstats.',
      event: '!cr',
      command: 'cr',
      enableKey: 'statsEnabled',
      messageKey: 'statsMessage',
      defaultMessage: '{player} · level {level} · {wins}W/{losses}L · {winrate}% WR · {crowns} three-crowns · {clan}',
      tokens: ['player', 'tag', 'level', 'wins', 'losses', 'draws', 'battles', 'winrate', 'crowns', 'challengemax', 'donations', 'totaldonations', 'clan', 'favcard'],
      previewSamples: {
        player: 'Bagel', tag: '#P2LQ0GR', level: '62', wins: '600', losses: '300', draws: '100',
        battles: '1000', winrate: '60', crowns: '120', challengemax: '12', donations: '50',
        totaldonations: '10000', clan: 'Bakery', favcard: 'Knight'
      }
    },
    {
      key: 'decks',
      label: '!cr decks',
      tagline: "The current battle deck with its average elixir (also !crdecks).",
      event: '!cr decks',
      command: 'cr decks',
      enableKey: 'decksEnabled',
      messageKey: 'decksMessage',
      defaultMessage: "{player}'s deck ({count}/8): {cards} · avg elixir {elixir}",
      tokens: ['player', 'tag', 'cards', 'support', 'elixir', 'count'],
      previewSamples: {
        player: 'Bagel', tag: '#P2LQ0GR',
        cards: 'Knight, Archers, Goblins, Giant, P.E.K.K.A, Minions, Fireball, Cannon',
        support: 'Tower Troop', elixir: '3.75', count: '8'
      }
    },
    {
      key: 'ranked',
      label: '!cr ranked',
      tagline: 'Path of Legends standing (also !crranked); legacy league seasons fill in for older records.',
      event: '!cr ranked',
      command: 'cr ranked',
      enableKey: 'rankedEnabled',
      messageKey: 'rankedMessage',
      defaultMessage: '{player} Path of Legends: league {league} · {trophies} trophies · rank #{rank} · best {besttrophies}',
      tokens: ['player', 'tag', 'league', 'trophies', 'rank', 'prevleague', 'prevtrophies', 'bestleague', 'besttrophies', 'bestrank'],
      previewSamples: {
        player: 'Bagel', tag: '#P2LQ0GR', league: '10', trophies: '2100', rank: '321',
        prevleague: '10', prevtrophies: '2050', bestleague: '10', besttrophies: '2400', bestrank: '42'
      }
    },
    {
      key: 'road',
      label: '!cr road',
      tagline: 'Trophy-road record and arena (also !crroad).',
      event: '!cr road',
      command: 'cr road',
      enableKey: 'roadEnabled',
      messageKey: 'roadMessage',
      defaultMessage: '{player}: {trophies} trophies · best {besttrophies} · {arena}',
      tokens: ['player', 'tag', 'trophies', 'besttrophies', 'arena'],
      previewSamples: { player: 'Bagel', tag: '#P2LQ0GR', trophies: '9123', besttrophies: '9345', arena: 'Legendary Arena' }
    }
  ],
  settings: [
    {
      key: 'account',
      label: 'Linked player tag',
      type: 'text',
      placeholder: '#P2LQ0GR',
      help: 'Default player for every command. Clash Royale has no name lookup, so this must be a player tag; leave blank only if you always type one.'
    }
  ]
};

File: app/twitch/sesame/modules/clashroyale.go
// Copyright (c) 2026 Adam Ousmer. All rights reserved.
// Proprietary. No license granted. See LICENSE.md.

package modules

import (
	"strings"
	"time"

	"ItsBagelBot/app/twitch/sesame/engine"
	"ItsBagelBot/app/twitch/sesame/module"
	gossiprpc "ItsBagelBot/internal/domain/rpc/gossip"
)

const clashroyaleModuleName = "clashroyale"

const clashroyaleCooldown = 10 * time.Second

const (
	defaultClashStatsTemplate  = "{player} · level {level} · {wins}W/{losses}L · {winrate}% WR · {crowns} three-crowns · {clan}"
	defaultClashDecksTemplate  = "{player}'s deck ({count}/8): {cards} · avg elixir {elixir}"
	defaultClashRankedTemplate = "{player} Path of Legends: league {league} · {trophies} trophies · rank #{rank} · best {besttrophies}"
	defaultClashRoadTemplate   = "{player}: {trophies} trophies · best {besttrophies} · {arena}"
)

const clashroyaleUnrankedText = "has no Path of Legends record this season"

type clashroyaleConfig struct {
	linkedAccountConfig

	StatsEnabled  string `json:"statsEnabled"`
	StatsMessage  string `json:"statsMessage"`
	DecksEnabled  string `json:"decksEnabled"`
	DecksMessage  string `json:"decksMessage"`
	RankedEnabled string `json:"rankedEnabled"`
	RankedMessage string `json:"rankedMessage"`
	RoadEnabled   string `json:"roadEnabled"`
	RoadMessage   string `json:"roadMessage"`
}

func ClashRoyale(d engine.Deps) module.Module {
	statsRun := externalCommand[clashroyaleConfig, gossiprpc.ClashRoyaleStatsReply]{
		route:      clashRoute("stats"),
		enabled:    func(c clashroyaleConfig) string { return c.StatsEnabled },
		message:    func(c clashroyaleConfig) string { return c.StatsMessage },
		fallback:   defaultClashStatsTemplate,
		tokens:     clashStatsTokens(),
		preferName: true,
	}.run(d)
	decksRun := externalCommand[clashroyaleConfig, gossiprpc.ClashRoyaleDecksReply]{
		route:      clashRoute("decks"),
		enabled:    func(c clashroyaleConfig) string { return c.DecksEnabled },
		message:    func(c clashroyaleConfig) string { return c.DecksMessage },
		fallback:   defaultClashDecksTemplate,
		tokens:     clashDecksTokens(),
		preferName: true,
	}.run(d)
	rankedRun := externalCommand[clashroyaleConfig, gossiprpc.ClashRoyaleRankedReply]{
		route:      clashRoute("ranked"),
		enabled:    func(c clashroyaleConfig) string { return c.RankedEnabled },
		message:    func(c clashroyaleConfig) string { return c.RankedMessage },
		fallback:   defaultClashRankedTemplate,
		tokens:     clashRankedTokens(),
		preferName: true,
		special:    clashRankedSpecial,
	}.run(d)
	roadRun := externalCommand[clashroyaleConfig, gossiprpc.ClashRoyaleTrophyRoadReply]{
		route:      clashRoute("trophy_road"),
		enabled:    func(c clashroyaleConfig) string { return c.RoadEnabled },
		message:    func(c clashroyaleConfig) string { return c.RoadMessage },
		fallback:   defaultClashRoadTemplate,
		tokens:     clashRoadTokens(),
		preferName: true,
	}.run(d)

	m := module.NewModule(clashroyaleModuleName, module.KindOptIn)
	m.Command("cr").Everyone().Cooldown(clashroyaleCooldown).
		Run(subDispatch(statsRun, map[string]module.RunFunc{
			"decks": decksRun, "deck": decksRun,
			"ranked": rankedRun, "pol": rankedRun,
			"road": roadRun, "trophy": roadRun, "trophies": roadRun,
		}))
	m.Command("crstats").Everyone().Cooldown(clashroyaleCooldown).Aliases("clashroyale").
		Run(statsRun)
	m.Command("crdecks").Everyone().Cooldown(clashroyaleCooldown).Aliases("crdeck").
		Run(decksRun)
	m.Command("crranked").Everyone().Cooldown(clashroyaleCooldown).Aliases("crpol").
		Run(rankedRun)
	m.Command("crroad").Everyone().Cooldown(clashroyaleCooldown).Aliases("crtrophy").
		Run(roadRun)
	return m.Build()
}

func clashRoute(endpoint string) engine.GossipRoute {
	return engine.GossipRoute{Provider: "clashroyale", Endpoint: endpoint}
}

func clashRankedSpecial(_ statsCall[clashroyaleConfig], r *gossiprpc.ClashRoyaleRankedReply) (string, bool) {
	if !r.Unranked {
		return "", false
	}
	return unrankedText(r.Player), true
}

func clashStatsTokens() module.TokenExpander[gossiprpc.ClashRoyaleStatsReply] {
	type reply = gossiprpc.ClashRoyaleStatsReply
	return module.TokenExpander[reply]{
		"player":         func(r *reply) string { return r.Player },
		"tag":            func(r *reply) string { return r.Tag },
		"level":          func(r *reply) string { return i64(int64(r.KingLevel)) },
		"wins":           func(r *reply) string { return i64(int64(r.Wins)) },
		"losses":         func(r *reply) string { return i64(int64(r.Losses)) },
		"draws":          func(r *reply) string { return i64(int64(r.Draws)) },
		"battles":        func(r *reply) string { return i64(int64(r.Battles)) },
		"winrate":        func(r *reply) string { return trimScore(r.WinRate) },
		"crowns":         func(r *reply) string { return i64(int64(r.ThreeCrownWins)) },
		"challengemax":   func(r *reply) string { return i64(int64(r.ChallengeMaxWins)) },
		"donations":      func(r *reply) string { return i64(int64(r.Donations)) },
		"totaldonations": func(r *reply) string { return i64(int64(r.TotalDonations)) },
		"clan": func(r *reply) string {
			if clan := r.Clan.Name; clan != "" {
				return clan
			}
			return "no clan"
		},
		"favcard": func(r *reply) string { return r.FavouriteCard.Name },
	}
}

func clashDecksTokens() module.TokenExpander[gossiprpc.ClashRoyaleDecksReply] {
	type reply = gossiprpc.ClashRoyaleDecksReply
	names := func(cards []gossiprpc.ClashRoyaleCard) string {
		parts := make([]string, len(cards))
		for i, card := range cards {
			parts[i] = card.Name
		}
		return strings.Join(parts, ", ")
	}
	return module.TokenExpander[reply]{
		"player":  func(r *reply) string { return r.Player },
		"tag":     func(r *reply) string { return r.Tag },
		"cards":   func(r *reply) string { return names(r.CurrentDeck) },
		"support": func(r *reply) string { return names(r.SupportCards) },
		"elixir":  func(r *reply) string { return trimScore(r.AverageElixir) },
		"count":   func(r *reply) string { return i64(int64(len(r.CurrentDeck))) },
	}
}

func clashRankedTokens() module.TokenExpander[gossiprpc.ClashRoyaleRankedReply] {
	type reply = gossiprpc.ClashRoyaleRankedReply
	return module.TokenExpander[reply]{
		"player":       func(r *reply) string { return r.Player },
		"tag":          func(r *reply) string { return r.Tag },
		"league":       func(r *reply) string { return i64(int64(r.Current.LeagueNumber)) },
		"trophies":     func(r *reply) string { return i64(int64(r.Current.Trophies)) },
		"rank":         func(r *reply) string { return i64(int64(r.Current.Rank)) },
		"prevleague":   func(r *reply) string { return i64(int64(r.Previous.LeagueNumber)) },
		"prevtrophies": func(r *reply) string { return i64(int64(r.Previous.Trophies)) },
		"bestleague":   func(r *reply) string { return i64(int64(r.Best.LeagueNumber)) },
		"besttrophies": func(r *reply) string { return i64(int64(r.Best.Trophies)) },
		"bestrank":     func(r *reply) string { return i64(int64(r.Best.Rank)) },
	}
}

func clashRoadTokens() module.TokenExpander[gossiprpc.ClashRoyaleTrophyRoadReply] {
	type reply = gossiprpc.ClashRoyaleTrophyRoadReply
	return module.TokenExpander[reply]{
		"player":       func(r *reply) string { return r.Player },
		"tag":          func(r *reply) string { return r.Tag },
		"trophies":     func(r *reply) string { return i64(int64(r.Trophies)) },
		"besttrophies": func(r *reply) string { return i64(int64(r.BestTrophies)) },
		"arena":        func(r *reply) string { return r.Arena.Name },
	}
}

func unrankedText(player string) string {
	return player + " " + clashroyaleUnrankedText
}
